Project manager: Add option to backup project when it will be changed

dsnopek opened the pull request:
The goal is to allow users to easily backup their projects if Godot will make irreversible changes to them. While this could make it more likely that folks will make backups in general, I'm mainly thinking about Android where managing files at all can be difficult (or require installing additional apps to make it not difficult).

Closes https://github.com/godotengine/godot-proposals/issues/2292

## Screenshots

### Opening an Old Project

If opening a project will lead to irreversible changes, it adds a "Backup project first" checkbox:

![Selection_317](https://github.com/user-attachments/assets/d0c2706b-fb71-4407-b0d6-e2b48d3eaa49)

If you check it and click "OK" (or "Full Conversion" if opening a Godot 3 project), then it'll take the user to a variation on the usual dialog for creating a project, in order to allow them to pick the directory and name the project:

![Selection_318](https://github.com/user-attachments/assets/8a185ebf-e3d4-45e6-b6e8-ddaf5853acbf)

After clicking "Duplicate", it will duplicate the project, but then continue on to open the original project.

### Duplicating a project directly

This PR also adds a "Duplicate" button to the right sidebar:

![Selection_315](https://github.com/user-attachments/assets/091f50ff-3f9a-40cf-8e1e-93ee343c9b03)

If the user selects a project, and then clicks this "Duplicate" button, it'll take the user to a variation on the usual dialog for creating a project similar to the one shown above:

![Selection_316](https://github.com/user-attachments/assets/44b30bb9-f276-415b-9918-2057436b6839)

However, this time it includes the "Edit Now" checkbox, which if checked, will lead to the newly duplicated project to be opened. If not, it only does the duplication, and then closes this dialog, returning the user to the list of projects.

## Caveats

- There is no progress dialog when copying the project: it just freezes until the copy is finished. However, it does the same thing when extracting a project from a ZIP file. We probably should have a progress dialog here for both installing from ZIP and project duplication, but since this matches pre-existing behavior, it seems like something that can be handled in a future PR

## Original List of Alternative Approaches

**UPDATE (2025-04-07):** The current PR now implements alternatives nr 1 and 2. I'm just keeping this section here so that early comments referring to these approaches still make sense :-)

1. Rather than sneaking in a "Backup Project" button into the dialog when changes are possible, we could have a "Copy Project" project button on the side and it'd be up to the user to use that button on their own. This takes the user out of the flow, so they may be less likely to do it, but it would be less weird that the project doesn't get opened immediately after completing the backup. Mockup:
    ![Selection_307](https://github.com/user-attachments/assets/21b0e041-1b84-4faf-9429-70adebcb3720)
2. Rather than being a button, there could be a "Make a backup first" checkbox in the dialog. Then regardless of which button the user clicks, it'll go to the dialog to make the backup, and then afterwards proceed to opening the project. Mockup:
    ![Selection_308](https://github.com/user-attachments/assets/b2cbf96d-b8ea-4793-88a3-2d21437a32be)
3. Rather than making a copy of the old project, and then opening the original (upgrading it), should this open the copy (leaving the original alone)? I was torn on which way to go, and I'm not sure I picked the right one. I guess if you think of the upgraded project as a continuation of the original, it's nice for that to be in the same directory going forward?

dsnopek committed:
Project manager: Add option to backup project when it will be changed

passivestar commented:
> we could have a "Copy Project" project button on the side

Duplicate button is useful on its own when making a lot of small similar projects regardless of whether there are other ways to backup. It would close https://github.com/godotengine/godot-proposals/issues/2292

Besides that dialog with a "Make Backup First" checkbox that makes a copy and then upgrades and opens the original makes the most sense to me

arkology commented:
I think using both (1) and (2) from alternative approaches would be great.

m4gr3d commented:
I'm also aligned with @passivestar and @arkology on going with (1) and (2) from the alternative approaches.

dsnopek commented:
I've updated the PR with alternative approaches nr 1 and 2, which means:

- Using a "Backup project first" checkbox (rather than an additional button)
- Adding a "Duplicate" button to the right sidebar, so users can duplicate arbitrary projects

I've updated the PR description with new screenshots showing all these changes!

Note: I looked at adding a progress dialog, but noticed that we are currently just freezing when installing a project from a ZIP file too, so doing the same for project duplication wouldn't be any worse. We probably should have a progress dialog here for both installing from ZIP and project duplication, but since this matches pre-existing behavior, it seems like something that can be handled in a future PR

So, taking this one out of DRAFT - it's ready for review :-)

devloglogan reviewed (approved):
Tested converting/duplicating some 4.x projects and it all worked as I would have expected. Code LGTM!

KoBeWi commented:
The duplicate option, or at least the backup one, should try to create the duplicate in the same directory as original. Right now it uses default project path.

dsnopek commented:
@KoBeWi Thanks for the review!

> The duplicate option, or at least the backup one, should try to create the duplicate in the same directory as original. Right now it uses default project path.

Added in my latest push!

> Project manager crashes if you try to duplicate a project that was removed externally while the manager was open

Eep! Thanks for catching that :-)

Fixed in my latest push

> Also I noticed that making project's backup/duplicate will change its name and directory name. Shouldn't they stay the same? The project name could be as is, the directory name could be `Original Directory Name (Copy)`, instead of using project's name like when creating new project.

Personally, I like being able to change the project name of the duplicate in that dialog.

Especially when doing a backup due to a version update, because you aren't going to open that project in the editor to rename it (otherwise it'd get updated) and this makes it easier to recognize in the project manager. But even when using the "Duplicate" button in the right sidebar, it prevents you from forgetting to change it when editing the project, avoiding future confusion in the project manager.

KoBeWi commented on editor/project_manager/project_dialog.cpp:
```diff
@@ -844,6 +874,20 @@ void ProjectDialog::show_dialog(bool p_reset_name) {
 
 			callable_mp((Control *)project_path, &Control::grab_focus).call_deferred();
+		} else if (mode == MODE_DUPLICATE) {
+			set_title(TTR("Duplicate Project"));
+			set_ok_button_text(TTR("Duplicate"));
```
```suggestion
			set_title(TTRC("Duplicate Project"));
			set_ok_button_text(TTRC("Duplicate"));
```

dsnopek replied:
Thanks! I've updated this (and all the `TTR()` to `TTRC()` suggestions) in my latest push

KoBeWi commented on editor/project_manager.cpp:
```diff
@@ -1518,6 +1583,11 @@ ProjectManager::ProjectManager() {
 
+			duplicate_btn = memnew(Button);
+			duplicate_btn->set_text(TTR("Duplicate"));
+			duplicate_btn->connect(SceneStringName(pressed), callable_mp(this, &ProjectManager::_duplicate_project));
```
You can do
```suggestion
			duplicate_btn->connect(SceneStringName(pressed), callable_mp(this, &ProjectManager::_duplicate_project_with_action).bind(POST_DUPLICATE_ACTION_NONE));
```
and remove `duplicate_project()` (or replace it).

KoBeWi commented on editor/project_manager.cpp:
```diff
@@ -1518,6 +1583,11 @@ ProjectManager::ProjectManager() {
 			project_list_sidebar->add_child(rename_btn);
 
+			duplicate_btn = memnew(Button);
+			duplicate_btn->set_text(TTR("Duplicate"));
```
```suggestion
			duplicate_btn->set_text(TTRC("Duplicate"));
```

KoBeWi commented on editor/project_manager.cpp:
```diff
@@ -1628,15 +1698,25 @@ ProjectManager::ProjectManager() {
+		ask_update_label->set_v_size_flags(SIZE_EXPAND_FILL);
+		ask_update_vb->add_child(ask_update_label);
+		ask_update_backup = memnew(CheckBox);
+		ask_update_backup->set_text(TTR("Backup project first"));
```
```suggestion
		ask_update_backup->set_text(TTRC("Backup project first"));
```

Repiteo commented:
Thanks!
